First draft of the RFP scope

ch1bo opened the pull request:
Done by @pgrange. I'm having a review on this now and will circulate for internal review as well.

pgrange committed:
CIP-52 RFP skeleton to fulfill

pgrange committed:
Collecting some stuff for the specification part

pgrange committed:
Auditors requirements

pgrange committed:
Section 2 Background

pgrange committed:
List artifacts involved in this audit

pgrange committed:
Add some tasks to be more described

pgrange committed:
Some more details

pgrange committed:
Point to stuff to grab and clean up below the line

ch1bo committed:
Jot down some thoughts on context, artifacts and tasks

pgrange committed:
Sorting some tasks for the auditor

pgrange committed:
Improve artifact schema

pgrange committed:
A bit of uniformity

pgrange committed:
Improve some structure

pgrange committed:
Clean useless comments

pgrange committed:
Hydra head v1 spec assessment

pgrange committed:
Stick to Google doc title of the spec name

pgrange committed:
Plutus script task

pgrange committed:
off-chain code assessment

pgrange committed:
Limit hypotheses for the environment

pgrange committed:
Head logic assessment

pgrange committed:
on-chain attacks

pgrange committed:
Network attacks

pgrange committed:
API assessment

pgrange committed:
Out of scope stuff

pgrange committed:
Refine hypotheses

ch1bo commented on security/RFP.md:
This means: we are connected to the cardano network at least once per contestation period?

pgrange replied:
Yes.
I'd like to stick to network hypotheses as formulated in the original paper or in head V1 but I must admit that, as of now, I can't manage to understand where this information is described in either of these documents :(

I found that in spec V1 but that does not look like what I'm looking for as it does not mention any time contraint:

> The head protocol gets correctly (and with completeness) notified about observed on-chain transactions belonging to the respective head instance.

pgrange replied:
We keep that one even if it's not perfect but better statement welcomed.

ch1bo commented on security/RFP.md:
Do you think we should describe this before the task? As a reader, I think I would already start thinking how to invalidate / comment on any of these and pay less attention to the tasks (which are the main part of the scope?)

ch1bo replied:
Also, I think this should rather be called "Assumptions". WDYT?

pgrange replied:
> As a reader, I think I would already start thinking how to invalidate / comment on any of these and pay less attention to the tasks

I don't understand your comment. Here we describe the assumptions under which the assessment must be performed. It means that any discovery which would not respect one on these assumptions would be rejected.

The idea here, is to avoid an auditor stating things like:

> If I have root access on the machine running the hydra-node and the cardano-node then I can do whatever I want.

So I don't expect an auditor to challenge any of these assumptions.

Having shared that, does that change how you understand this section? If so, would you have any improvement to suggest for the next reader?

ch1bo commented on security/RFP.md:
phew.. not sure if this is so easy / possible using that test suite

pgrange replied:
I feel like, as an auditor, I would be super happy to see that kind of tests to help me build a conviction that the code does what it should. Agree that it's probably not trivial to read but, anyway, the code is probably not trivial to read for an external auditor, right?

ch1bo commented on security/RFP.md:
It think this overlaps a lot with the next task. Shall we merge them?

pgrange replied:
I don't see both tasks as overlapping and don't feel like we should merge them.

The next task, _Hydra plutus scripts are immune to common Cardano smart-contract weaknesses_, looks to me quite generic and could apply to any plutus scripts.

When this task, _Hydra plutus scripts are consistent with Hydra Head v1 specification_ is really about understanding the specification and checking that we apply the same validators in our plutus scripts than in the spec V1.

For instance, we can correctly implement the specification in our validator but be subject to some [hard limits](https://plutus--4604.org.readthedocs.build/en/4604/reference/common-weaknesses/hard-limits.html). That's why I think it has value to identify both task and emphasis on _we don't want you to just check we implemented the spec_ but also something like _the validators can actually run on cardano_.

ch1bo commented on security/RFP.md:
This is repeated from above. Overall I think the RFP/scope contains a lot of repetition. It's tiring to read.

pgrange replied:
I resolve this one because I feel like you already fixed that. If I'm wrong, please re-open.

ch1bo commented on security/RFP.md:
Can't we do that on the specification itself? If we audit the spec and it's fine, we "only" need to audit the chain layer to be consistent with the assumptions from the paper?

Also applies to the network task below.

pgrange replied:
Good point. Maybe I should add something for the spec V1 audit part about that.

I feel like, however, that it could be interesting to still check: is our actual code immune to actual on chain invalid transactions?

I don't know 🤔

ch1bo commented on security/RFP.md:
I think this task can be about network DoS. Are you considering asking for checking whether the hydra network connections can be used to DoS the whole machine, which would affect the network connection of the `cardano-node` on the same machine? If yes .. that is likely possible and a valid point for this task!

ch1bo commented on security/RFP.md:
I don't think we should be using the same assumptions for the implementation. The specification defines a more narrow space to reason about some things. If we consider the implementation we cannot simply assume (for example) "all network connections are authenticated and never fail".

ch1bo committed:
Rewrite some things in the rfp scope

ch1bo reviewed (commented):
I think we have collected enough material and I think the current document is very long and repetitive. What do you think about:
* Just give an overview in the context section and inlining artifact descriptions within the corresponding tasks? It seems like we would want to point to specific modules / documentation anyways and the mapping is mostly 1:1
* Tasks 2/3 are very much overlapping. Shall we just merge them?
* Tasks 6/7/8 are very similar as well (indicated by the same text passages). Can we merge them? It's basically checking that none of the channels two channels **into our security domain** can be used to do X.

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+
+### Hydra Head protocol implementation faithfully reflects the head state through the API
+
+Given the assumptions above, the API is only accessible locally from the trusted machine running the `hydra-node`. Moreover, we don't consider the client API as an adversarial party so any attack through the API is considered out of scope. However, it is important that a running node does reflect the current stat of a head faithfully to its API client(s).
```
I would say that we trust the client as much as we trust the hydra-node. The way I see it:

As a hydra head user,
I own a trusted machine,
On which I run a trusted instance of a hydra-node,
To which I connect my trusted client application
And I want to know if somebody else than me can do nasty things to me.

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+* The cardano-node software is trusted;
+* The hydra-node software is trusted;
+* The hydra-node API is only accessible from the local machine;
+* For any client software connecting to the hydra-node through its API, the hydra-node and the client software both run on the same machine;
```
We should explicit that what is not listed as truster is not trusted

pgrange replied:
We should explicit that network and chain are the two entry points for attacks

pgrange replied:
(Add Charles on the picture ?)

pgrange replied:
I would consider the client as trusted (see other discussion below)

pgrange replied:
However it could make sense to not fully trust the client depending on the use case we can imagine. Do we want that in this version of this audit?

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+* The cardano-node and the hydra-node software communicate through local unix socket;
+* The cardano-node software is trusted;
+* The hydra-node software is trusted;
+* The hydra-node API is only accessible from the local machine;
```
We should state that with a bit more of generality and not speak about _local machine_ but maybe more about secured connections.

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+* The local machine running a hydra-node is trusted;
+* The hydra and cardano signing keys storage and operation are trusted.
+* For any legitimate hydra-node, the network is resilient enough that it can synchronize the cardano blockchain at regular interval smaller than any head contestation period they participate in.
+
```
About the network, we should double check what is in the original paper

pgrange replied:
About the network, we keep that one even if it's not perfect but better statement welcomed.

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+* identification of network attack that would lead to a denial of service for the hydra node
+* identification of network attack that would invalidate one of the security properties of the Coordinated Hydra Head V1 specification.
+
+TODO: should we explicitly state that we are not interested in network denial of service attack? In the assumptions, should we consider only that locally submitted transaction will end up reaching the cardano network or should we enforce stronger network propoerties? For instance, if a node totally loose its network connection, is it OK? For how long is it OK?
```
About the network, we should double check what is in the original paper

pgrange replied:
Replaced the TODO with this:
```
Generic denial of service attack against the node itself is out of scope of this audit. One should only focus on specific hydra network layer attacks that could generate a Denial of Service because of an issue in the hydra-node itself.
```

pgrange replied:
Also, in the assumptions, we ruled out attack that would imply fully isolating the node from the network. Because, of course, this would prevent the node from seeing a malicious close transaction and contest it. But... that's an obvious one we don't want an auditor to waste time on.

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+
+We are requesting proposals to assess the following statements, which will be detailed in the next sections. We will also accept proposals covering subsets of this list, but would like to get at least the highest priority tasks covered. The responder should briefly explain how they intend to address each task. If not all tasks are covered, individual estimation of efforts is required.
+
+-- Specification
```
Specified later but we can give some insight upfront 👍

pgrange commented on security/RFP.md:
```diff
@@ -0,0 +1,241 @@
+
+TODO: split the RFP?
+
+### Coordinated Hydra Head V1 specification is consistent with the original publication
```
You're right, the proofs for the spec v1 is being written as we speak so we should remove that item 👍

pgrange replied:
Maybe we want the auditor to assess spec v1 proofs

pgrange committed:
Addressing several comments

pgrange committed:
Improve network hypothesys

pgrange committed:
Moving the boring artifacts section and the end of the document

Usefull information as a reference but boring at this point in time.

pgrange committed:
Uniformity with artifacts name

pgrange committed:
FIX tyop

pgrange committed:
OUT OF SCOPE cleanup

pgrange committed:
Cleanup FIXME

pgrange committed:
Clarify scope

We want dynamic analysis.

The organization managing the keys is trusted but shall one find a
workaround through our software to get the keys, we would be interested.

The hydra-node software is not truster as it is the thing under
scrutiny. The system running it is, though.

pgrange committed:
Clarify plutus audit

Everybody ask why not merge both section so... let's merge them.

Also, it's a security audit so we don't ask for optimization suggestions
as this would be out of scope.

pgrange committed:
FIX TODOs

pgrange committed:
FIX tyops

pgrange commented:
I've included this text in an internal document to prepare for an official RFP.

I recommend we do not merge this P.R.

I'd like to keep it open though, as a draft, if it's ok. My idea is to, at some point, transform this text into documentation on hydra.family about what has been assessed or what will be assessed or what we would recommend someone to assess. Not sure yet.

What do you think?

pgrange commented:
Won't merge as valuable output has been captured in #606
